pallets/bridge/src/lib.rs

```diff
@@ -0,0 +1,621 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+pub mod hashing;
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use codec::{Decode, Encode, EncodeLike};
+	pub use frame_support::{
+		pallet_prelude::*, traits::StorageVersion, weights::GetDispatchInfo, PalletId, Parameter,
+	};
+	use frame_system::{
+		pallet_prelude::*,
+		{self as system},
+	};
+	use scale_info::TypeInfo;
+	pub use sp_core::U256;
+	use sp_runtime::{
+		traits::{AccountIdConversion, Dispatchable},
+		RuntimeDebug,
+	};
+	use sp_std::prelude::*;
+
+	const DEFAULT_RELAYER_THRESHOLD: u32 = 1;
+	const MODULE_ID: PalletId = PalletId(*b"litry/bg");
+
+	pub type BridgeChainId = u8;
+	pub type DepositNonce = u64;
+	pub type ResourceId = [u8; 32];
+
+	/// Helper function to concatenate a chain ID and some bytes to produce a resource ID.
+	/// The common format is (31 bytes unique ID + 1 byte chain ID).
+	pub fn derive_resource_id(chain: u8, id: &[u8]) -> ResourceId {
+		let mut r_id: ResourceId = [0; 32];
+		r_id[31] = chain; // last byte is chain id
+		let range = if id.len() > 31 { 31 } else { id.len() }; // Use at most 31 bytes
+		for i in 0..range {
+			r_id[30 - i] = id[range - 1 - i]; // Ensure left padding for eth compatibility
+		}
+		r_id
+	}
+
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum ProposalStatus {
+		Initiated,
+		Approved,
+		Rejected,
+	}
+
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub struct ProposalVotes<AccountId, BlockNumber> {
+		pub votes_for: Vec<AccountId>,
+		pub votes_against: Vec<AccountId>,
+		pub status: ProposalStatus,
+		pub expiry: BlockNumber,
+	}
+
+	#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+	pub enum BridgeEvent {
+		FungibleTransfer(BridgeChainId, DepositNonce, ResourceId, U256, Vec<u8>),
+		NonFungibleTransfer(BridgeChainId, DepositNonce, ResourceId, Vec<u8>, Vec<u8>, Vec<u8>),
+		GenericTransfer(BridgeChainId, DepositNonce, ResourceId, Vec<u8>),
+	}
+
+	impl<A: PartialEq, B: PartialOrd + Default> ProposalVotes<A, B> {
+		/// Attempts to mark the proposal as approve or rejected.
+		/// Returns true if the status changes from active.
+		pub fn try_to_complete(&mut self, threshold: u32, total: u32) -> ProposalStatus {
+			if self.votes_for.len() >= threshold as usize {
+				self.status = ProposalStatus::Approved;
+				ProposalStatus::Approved
+			} else if total >= threshold && self.votes_against.len() as u32 + threshold > total {
+				self.status = ProposalStatus::Rejected;
+				ProposalStatus::Rejected
+			} else {
+				ProposalStatus::Initiated
+			}
+		}
+
+		/// Returns true if the proposal has been rejected or approved, otherwise false.
+		fn is_complete(&self) -> bool {
+			self.status != ProposalStatus::Initiated
+		}
+
+		/// Returns true if `who` has voted for or against the proposal
+		fn has_voted(&self, who: &A) -> bool {
```

Comment:
What happens if a voter votes for both `for` and `against`? Do we need to check this?

Reply:
yes, the vote is finalized with the assumption each one just vote once, for or against.